test/geocoders/baidu.py

```diff
@@ -5,16 +5,34 @@
 from geopy.point import Point
 from geopy.geocoders import Baidu
 from test.geocoders.util import GeocoderTestBase, env
+from geopy.exc import GeocoderAuthenticationFailure
+from geopy.compat import py3k
 
 
 class BaiduTestCaseUnitTest(GeocoderTestBase):
 
-    def test_user_agent_custom(self):
-        geocoder = Baidu(
+    @classmethod
+    def setUpClass(cls):
+        cls.geocoder = Baidu(
             api_key='DUMMYKEY1234',
             user_agent='my_user_agent/1.0'
         )
-        self.assertEqual(geocoder.headers['User-Agent'], 'my_user_agent/1.0')
+
+    def test_user_agent_custom(self):
+        self.assertEqual(self.geocoder.headers['User-Agent'], 'my_user_agent/1.0')
+
+    def test_invalid_ak(self):
+        if py3k:
+            assert_raise_msg = self.assertRaisesRegex
+        else:
+            assert_raise_msg = self.assertRaisesRegexp
+
+        with assert_raise_msg(GeocoderAuthenticationFailure, 'Invalid AK'):
```

Comment:
Another way to do the same without tinkering around compatibility:
```
    with self.assertRaises(GeocoderAuthenticationFailure) as cm:
        ...
    self.assertEqual(str(cm.exception), 'Invalid AK')
```

To me it looks better than `assertRaisesRegex`. What do you think?

Also I'm wondering: does the `self.reverse_run` command ever run? I feel that `self.geocode_run` causes an exception being thrown, and the `self.reverse_run` call is never tested. To fix that, each call should be wrapped in their own `assertRaises`/`assertRaisesRegex` context.

Reply:
Thanks, `self.reverse_run` cannot be reached. I will remove this line afterwards.